### PR TITLE
add support override for csi-node-driver

### PR DESCRIPTION
**Pull request by zoezhangmattr**
## Description
try to support overrides for csi-node-driver daemonset including tolerations
see this issue https://github.com/projectcalico/calico/issues/7070

## For PR author

- [x] Tests for change.
- [x] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: api/v1/csi_node_driver.go
Comment (Josh-Tigera) [inline api/v1/csi_node_driver.go:100]: Please add an implementation for this function.

File: api/v1/csi_node_driver.go
Comment (Josh-Tigera) [inline api/v1/csi_node_driver.go:83]: Please add an implementation for this function.

File: api/v1/csi_node_driver.go
Comment (Josh-Tigera) [inline api/v1/csi_node_driver.go:96]: Please add a comment here indicating that the CSI DaemonSet does not contain any init containers so that we know the lack of implementation was deliberate.

**Review by Josh-Tigera — CHANGES_REQUESTED** [review]
Thanks for getting this up so quickly. A few minor changes I'd like to see but otherwise looking good.

File: api/v1/csi_node_driver.go
Comment (zoezhangmattr) [reply api/v1/csi_node_driver.go:130]: thank you for review, applied

File: api/v1/csi_node_driver.go
Comment (zoezhangmattr) [reply api/v1/csi_node_driver.go:109]: thank you for review, applied

File: api/v1/csi_node_driver.go
Comment (zoezhangmattr) [reply api/v1/csi_node_driver.go:125]: thank you for review, applied

File: api/v1/csi_node_driver.go
Comment (zoezhangmattr) [reply api/v1/csi_node_driver.go:109]: > Please add an implementation for this function.

thank you for review, applied

File: api/v1/csi_node_driver.go
Comment (zoezhangmattr) [reply api/v1/csi_node_driver.go:130]: > Please add an implementation for this function.

thank you for review, applied

File: api/v1/csi_node_driver.go
Comment (zoezhangmattr) [reply api/v1/csi_node_driver.go:125]: > Please add a comment here indicating that the CSI DaemonSet does not contain any init containers so that we know the lack of implementation was deliberate.

thank you for review, applied

**Comment by Josh-Tigera** [comment]
@zoezhangmattr I've requested someone from @tigera/tigera-operator-maintainers to also have a look since we ultimately need one of them to sign off before we can merge.

**Review by tmjd — APPROVED** [review]
LGTM

**Review by tmjd — CHANGES_REQUESTED** [review]
Thanks for this PR.
Looks like your branch needs to be updated with latest master.
You also need a corresponding update to https://github.com/tigera/operator/blob/master/pkg/controller/utils/merge.go#L149 to add this new new field and type to the merge logic.

**Comment by zoezhangmattr** [comment]
> Thanks for this PR. Looks like your branch needs to be updated with latest master. You also need a corresponding update to https://github.com/tigera/operator/blob/master/pkg/controller/utils/merge.go#L149 to add this new new field and type to the merge logic.

thanks for review, i will update it!

**Comment by caseydavenport** [comment]
@Josh-Tigera @tmjd looks like this needs a re-approval from you guys

**Review by tmjd — APPROVED** [review]
LGTM, I kicked off CI again since the last run failed. I'm not sure if it needs another update from master since it has sat for a while again. Though github doesn't report the branch as out of date.

**Comment by tmjd** [comment]
I went ahead and merged this after pulling down the branch and making sure the CI failure was only because of master having some additional updates.
Thank you for submitting this @zoezhangmattr

**Comment by zoezhangmattr** [comment]
> I went ahead and merged this after pulling down the branch and making sure the CI failure was only because of master having some additional updates. Thank you for submitting this @zoezhangmattr

thank you for your reivew , 
and @Josh-Tigera for the guide to do the change. :)

**Comment by Josh-Tigera** [comment]
@zoezhangmattr any time, thanks for the contribution!

**Comment by tmjd** [comment]
There is an issue with this https://github.com/projectcalico/calico/issues/8168 and I've put up a fix here https://github.com/tigera/operator/pull/2977

**Pull request merged**
